quarkus-kubernetes java.lang.ClassNotFoundException: io.fabric8.kubernetes.api.model.GenericKubernetesResourceBuilder

### Describe the bug

The extension `quarkus-kubernetes` fails if I add a custom resource in `src/main/kubernetes/kubernetes.yml` with :

```
[INFO] --- quarkus-maven-plugin:2.1.2.Final:build (default) @ my ---
[INFO] [org.jboss.threads] JBoss Threads version 3.4.2.Final
[INFO] Checking for existing resources in: /Users/anthonydahanne/workspaces/my/src/main/kubernetes.
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  9.150 s
[INFO] Finished at: 2021-08-12T13:30:53-04:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal io.quarkus.platform:quarkus-maven-plugin:2.1.2.Final:build (default) on project my: Failed to build quarkus application: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
[ERROR]         [error]: Build step io.quarkus.kubernetes.deployment.KubernetesProcessor#build threw an exception: java.lang.IllegalStateException: Failed to create builder for: class io.fabric8.kubernetes.api.model.GenericKubernetesResource
[ERROR]         at io.fabric8.kubernetes.api.builder.BaseFluent.builderOf(BaseFluent.java:30)
[ERROR]         at io.fabric8.kubernetes.api.model.BaseKubernetesListFluentImpl.addToItems(BaseKubernetesListFluentImpl.java:134)
[ERROR]         at io.fabric8.kubernetes.api.model.BaseKubernetesListFluentImpl.withItems(BaseKubernetesListFluentImpl.java:254)
[ERROR]         at io.dekorate.utils.Serialization.unmarshalAsList(Serialization.java:196)
[ERROR]         at io.dekorate.processor.SimpleFileReader.readYml(SimpleFileReader.java:78)
[ERROR]         at java.base/java.util.stream.Collectors.lambda$uniqKeysMapAccumulator$1(Collectors.java:180)
[ERROR]         at java.base/java.util.stream.ReduceOps$3ReducingSink.accept(ReduceOps.java:169)
[ERROR]         at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:179)
[ERROR]         at java.base/java.util.HashMap$KeySpliterator.forEachRemaining(HashMap.java:1694)
[ERROR]         at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
[ERROR]         at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
[ERROR]         at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
[ERROR]         at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
[ERROR]         at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682)
[ERROR]         at io.dekorate.processor.SimpleFileReader.findApplicableResources(SimpleFileReader.java:101)
[ERROR]         at io.dekorate.processor.SimpleFileReader.read(SimpleFileReader.java:52)
[ERROR]         at io.dekorate.Session.lambda$readExistingResources$6(Session.java:291)
[ERROR]         at java.base/java.util.Optional.ifPresent(Optional.java:178)
[ERROR]         at io.dekorate.Session.readExistingResources(Session.java:291)
[ERROR]         at io.dekorate.Session.generate(Session.java:281)
[ERROR]         at io.dekorate.Session.close(Session.java:248)
[ERROR]         at io.quarkus.kubernetes.deployment.KubernetesProcessor.lambda$build$4(KubernetesProcessor.java:165)
[ERROR]         at java.base/java.util.Optional.ifPresent(Optional.java:178)
[ERROR]         at io.quarkus.kubernetes.deployment.KubernetesProcessor.build(KubernetesProcessor.java:123)
[ERROR]         at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[ERROR]         at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:78)
[ERROR]         at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[ERROR]         at java.base/java.lang.reflect.Method.invoke(Method.java:567)
[ERROR]         at io.quarkus.deployment.ExtensionLoader$2.execute(ExtensionLoader.java:820)
[ERROR]         at io.quarkus.builder.BuildContext.run(BuildContext.java:277)
[ERROR]         at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
[ERROR]         at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
[ERROR]         at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
[ERROR]         at java.base/java.lang.Thread.run(Thread.java:831)
[ERROR]         at org.jboss.threads.JBossThread.run(JBossThread.java:501)
[ERROR] Caused by: java.lang.ClassNotFoundException: io.fabric8.kubernetes.api.model.GenericKubernetesResourceBuilder
[ERROR]         at org.codehaus.plexus.classworlds.strategy.SelfFirstStrategy.loadClass(SelfFirstStrategy.java:50)
[ERROR]         at org.codehaus.plexus.classworlds.realm.ClassRealm.unsynchronizedLoadClass(ClassRealm.java:271)
[ERROR]         at org.codehaus.plexus.classworlds.realm.ClassRealm.loadClass(ClassRealm.java:247)
[ERROR]         at org.codehaus.plexus.classworlds.realm.ClassRealm.loadClass(ClassRealm.java:239)
[ERROR]         at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:455)
[ERROR]         at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:405)
[ERROR]         at java.base/java.lang.Class.forName0(Native Method)
[ERROR]         at java.base/java.lang.Class.forName(Class.java:375)
[ERROR]         at io.fabric8.kubernetes.api.builder.BaseFluent.builderOf(BaseFluent.java:27)
[ERROR]         ... 34 more
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.

```


### Expected behavior

If I save this manifest in `src/main/kubernetes/kubernetes.yml`, the extension should concatenate this additionnal Kubernetes resource to the file located at: `target/kubernetes/kubernetes.yml`


### Actual behavior

If I save this manifest in `src/main/kubernetes/kubernetes.yml`, the extension will consider it and then fails


### How to Reproduce?


Using the extension:

     <dependency>
      <groupId>io.quarkus</groupId>
      <artifactId>quarkus-kubernetes</artifactId>
    </dependency>

I was able to successfully generate  `Service` and `Deployment`  resources in a file located in `target/kubernetes/kubernetes.yml`

In my deployment environment, I also need to an Istio resource, that's looking like this:

```
---
apiVersion: networking.istio.io/v1beta1
kind: VirtualService
metadata:
  name: my-vs
spec:
  gateways:
    - my-default-ingressgw
  hosts:
    - my.domain.com
  http:
    - match:
        - uri:
            prefix: /
      route:
        - destination:
            host: my
            port:
              number: 80

```

### Output of `uname -a` or `ver`

Darwin host@employer.com 19.6.0 Darwin Kernel Version 19.6.0: Thu May  6 00:48:39 PDT 2021; root:xnu-6153.141.33~1/RELEASE_X86_64 x86_64

### Output of `java -version`

openjdk version "16.0.1" 2021-04-20 OpenJDK Runtime Environment AdoptOpenJDK-16.0.1+9 (build 16.0.1+9) OpenJDK 64-Bit Server VM AdoptOpenJDK-16.0.1+9 (build 16.0.1+9, mixed mode, sharing)

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

2.1.2.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.8.1 (05c21c65bdfed0f71a2f2ada8b84da59348c4c5d)

### Additional information

_No response_

--- a/extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/KubernetesDeployer.java
+++ b/extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/KubernetesDeployer.java
@@ -24,12 +24,16 @@ import org.jboss.logging.Logger;
 
 import io.dekorate.utils.Clients;
 import io.dekorate.utils.Serialization;
+import io.fabric8.kubernetes.api.model.APIResourceList;
+import io.fabric8.kubernetes.api.model.GenericKubernetesResource;
 import io.fabric8.kubernetes.api.model.HasMetadata;
 import io.fabric8.kubernetes.api.model.KubernetesList;
 import io.fabric8.kubernetes.api.model.Service;
 import io.fabric8.kubernetes.client.Config;
 import io.fabric8.kubernetes.client.KubernetesClient;
 import io.fabric8.kubernetes.client.KubernetesClientException;
+import io.fabric8.kubernetes.client.dsl.base.ResourceDefinitionContext;
+import io.fabric8.kubernetes.client.utils.ApiVersionUtil;
 import io.fabric8.openshift.api.model.Route;
 import io.fabric8.openshift.client.OpenShiftClient;
 import io.quarkus.container.image.deployment.ContainerImageCapabilitiesUtil;
@@ -191,22 +195,41 @@ public class KubernetesDeployer {
         try (FileInputStream fis = new FileInputStream(manifest)) {
             KubernetesList list = Serialization.unmarshalAsList(fis);
             list.getItems().stream().filter(distinctByResourceKey()).forEach(i -> {
-                final var r = client.resource(i).inNamespace(namespace);
-                if (shouldDeleteExisting(deploymentTarget, i)) {
-                    r.delete();
-                    r.waitUntilCondition(Objects::isNull, 10, TimeUnit.SECONDS);
-                }
-                try {
-                    r.createOrReplace();
-                } catch (Exception e) {
-                    if (e instanceof InterruptedException) {
-                        throw e;
-                    } else if (isOptional(optionalResourceDefinitions, i)) {
-                        log.warn("Failed to apply: " + i.getKind() + " " + i.getMetadata().getName()
-                                + ", possilby due to missing a CRD apiVersion: " + i.getApiVersion() + " and kind: "
-                                + i.getKind() + ".");
-                    } else {
-                        throw e;
+                if (i instanceof GenericKubernetesResource) {
+                    GenericKubernetesResource genericResource = (GenericKubernetesResource) i;
+                    ResourceDefinitionContext context = getGenericResourceContext(client, genericResource)
+                            .orElseThrow(() -> new IllegalStateException("Could not retrieve API resource information for:"
+                                    + i.getApiVersion() + " " + i.getKind() + ". Is the CRD for the resource available?"));
+
+                    client.genericKubernetesResources(context).withName(i.getMetadata().getName())
+                            .createOrReplace(genericResource);
+                } else {
+                    final var r = client.resource(i);
+                    if (shouldDeleteExisting(deploymentTarget, i)) {
+                        r.delete();
+                        try {
+                            r.waitUntilCondition(Objects::isNull, 10, TimeUnit.SECONDS);
+                        } catch (Exception e) {
+                            if (e instanceof InterruptedException) {
+                                throw e;
+                            }
+                            //This is something that should not really happen. it's not a fatal condition so let's just log.
+                            log.warn("Failed to wait for the deletion of: " + i.getApiVersion() + " " + i.getKind() + " "
+                                    + i.getMetadata().getName() + ". Is the resource waitable?");
+                        }
+                    }
+                    try {
+                        r.createOrReplace();
+                    } catch (Exception e) {
+                        if (e instanceof InterruptedException) {
+                            throw e;
+                        } else if (isOptional(optionalResourceDefinitions, i)) {
+                            log.warn("Failed to apply: " + i.getKind() + " " + i.getMetadata().getName()
+                                    + ", possilby due to missing a CRD apiVersion: " + i.getApiVersion() + " and kind: "
+                                    + i.getKind() + ".");
+                        } else {
+                            throw e;
+                        }
                     }
                 }
                 log.info("Applied: " + i.getKind() + " " + i.getMetadata().getName() + ".");
@@ -248,6 +271,32 @@ public class KubernetesDeployer {
                 break;
             }
         }
+    }
+
+    /**
+     * Obtain everything the APIResourceList from the server and extract all the info we need in order to know how to create /
+     * delete the specified generic resource.
+     * 
+     * @param client the client instance to use to query the server.
+     * @param resource the generic resource.
+     * @return an optional {@link ResourceDefinitionContext} with the resource info or empty if resource could not be matched.
+     */
+    private static Optional<ResourceDefinitionContext> getGenericResourceContext(KubernetesClient client,
+            GenericKubernetesResource resource) {
+        APIResourceList apiResourceList = client.getApiResources(resource.getApiVersion());
+        if (apiResourceList == null || apiResourceList.getResources() == null || apiResourceList.getResources().isEmpty()) {
+            return Optional.empty();
+        }
+        return client.getApiResources(resource.getApiVersion()).getResources().stream()
+                .filter(r -> r.getKind().equals(resource.getKind()))
+                .map(r -> new ResourceDefinitionContext.Builder()
+                        .withGroup(ApiVersionUtil.trimGroup(resource.getApiVersion()))
+                        .withVersion(ApiVersionUtil.trimVersion(resource.getApiVersion()))
+                        .withKind(r.getKind())
+                        .withNamespaced(r.getNamespaced())
+                        .withPlural(r.getName())
+                        .build())
+                .findFirst();
     }
 
     private static boolean isOptional(List<KubernetesOptionalResourceDefinitionBuildItem> optionalResourceDefinitions,

